src/devices/Pn5180/Pn5180.cs

```diff
@@ -0,0 +1,1496 @@
+                    if (toReceive[1] != targetNumber)
+                    {
+                        return -1;
+                    }
+
+                    // Is it a valid packet?
+                    if (toReceive[0] == (0b0000_1010 | (card.LastBlockMark ? 1 : 0)))
+                    {
+                        toReceive.Slice(2).CopyTo(dataFromCard);
+                        dataFromCard[numBytes - 2] = 0x00;
+                        card.LastBlockMark = !card.LastBlockMark;
+                        return numBytes - 1;
+                    }
+
+                    // Case of a chained packet, we need to send acknowledgment and read more data
+                    if (toReceive[0] == (0b0001_1010 | (card.LastBlockMark ? 1 : 0)))
+                    {
+                        // Copy what we have already
+                        toReceive.Slice(2, numBytes - 2).CopyTo(dataFromCard);
+                        // Send Ack with the right mark
+                        card.LastBlockMark = !card.LastBlockMark;
+                        Span<byte> rBlockChain = new byte[2] { (byte)(0b1010_1010 | (card.LastBlockMark ? 1 : 0)), targetNumber };
+                        var numBytes2 = TranceiveTypeB(card, rBlockChain, toReceive);
+
+                        if (numBytes2 >= 2)
+                        {
+                            // If not the right target, then not for us
+                            if (!(toReceive[1] == targetNumber))
+                            {
+                                return -1;
+                            }
+
+                            if (!(toReceive[0] == (0b0000_1010 | (card.LastBlockMark ? 1 : 0))))
+                            {
+                                return -1;
+                            }
+
+                            toReceive.Slice(2, numBytes2 - 2).CopyTo(dataFromCard.Slice(numBytes - 2));
+                            dataFromCard[numBytes - 2 + numBytes2 - 2] = 0x00;
+                            card.LastBlockMark = !card.LastBlockMark;
+                            return numBytes - 2 + numBytes2 - 1;
+                        }
+                    }
+
+                    // Is it an extension time block request block?
+                    if (toReceive[0] == 0b1111_1010)
+                    {
+                        // Agree and send the same
+                        Span<byte> sBlock = new byte[] { 0b1111_1010, card.Card.TargetNumber, toReceive[2] };
+                        var ret = SendDataToCard(sBlock);
+                        if (maxTries++ == MaxTries)
+                        {
+                            return -1;
+                        }
+
+                        numBytes = ReadWithTimeout(dataFromCard, (int)(toReceive[2] * card.Card.FrameWaitingTime / 1000));
+                        goto IBlock;
+                    }
+                }
+
+                return numBytes;
+            }
+            else
```

Comment:
I'm not sure where this `if .. else` start but you might want to swap the condition there so else is first (easier to read)

Reply:
ho yes, done!